model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt

```diff
@@ -178,53 +199,76 @@ class RepositoriesManager(val client: LocalModelClient) {
         }
     }
 
-    fun removeBranches(repository: RepositoryId, branchNames: Set<String>) {
+    suspend fun removeBranches(repository: RepositoryId, branchNames: Set<String>) {
+        return store.runTransactionSuspendable {
+            removeBranchesInsideTransaction(repository, branchNames)
+        }
+    }
+
+    /**
+     * Same as [removeBranchesInsideTransaction] but assumes an already running transaction.
+     */
+    fun removeBranchesInsideTransaction(repository: RepositoryId, branchNames: Set<String>) {
         if (branchNames.isEmpty()) return
-        store.runTransaction {
-            val key = branchListKey(repository)
-            val existingBranches = store[key]?.lines()?.toSet() ?: emptySet()
-            val remainingBranches = existingBranches - branchNames
-            store.put(key, remainingBranches.joinToString("\n"))
-            for (branchName in branchNames) {
-                putVersionHash(repository.getBranchReference(branchName), null)
-            }
+        val key = branchListKey(repository)
+        val existingBranches = store[key]?.lines()?.toSet() ?: emptySet()
+        val remainingBranches = existingBranches - branchNames
+        store.put(key, remainingBranches.joinToString("\n"))
+        for (branchName in branchNames) {
+            putVersionHash(repository.getBranchReference(branchName), null)
         }
     }
 
-    fun mergeChanges(branch: BranchReference, newVersionHash: String): String {
-        var result: String? = null
-        store.runTransaction {
-            val branchKey = branchKey(branch)
-            val headHash = getVersionHash(branch)
-            val mergedHash = if (headHash == null) {
-                newVersionHash
-            } else {
-                val headVersion = CLVersion(headHash, client.storeCache)
-                val newVersion = CLVersion(newVersionHash, client.storeCache)
-                require(headVersion.tree.getId() == newVersion.tree.getId()) {
-                    "Attempt to merge a model with ID '${newVersion.tree.getId()}'" +
-                        " into one with ID '${headVersion.tree.getId()}'"
-                }
-                val mergedVersion = VersionMerger(client.storeCache, client.idGenerator)
-                    .mergeChange(headVersion, newVersion)
-                mergedVersion.hash
+    suspend fun mergeChanges(branch: BranchReference, newVersionHash: String): String {
```

Comment:
Having a suspendable and non-suspendable version of each function everywhere is not a nice solution. Transactions and coroutines don't seem mix very well.

I don't have a better solution yet, so let's merge it for now and change it in case we find a better alternative.

Reply:
If we could create a transaction that is not bound to a thread in Ingite, we could store it in the
coroutine context.
I got hung up on another issue when trying to look into Ignite code.
But in general, this should be doable.